docs/ADR-8-Use-RIO-in-cardano‐cli.md

```diff
@@ -0,0 +1,129 @@
+# Status
+- [ ] Adopted YYYY/MM/DD
+
+# Context
+- In `cardano-cli` We have the (anti-)pattern `ExceptT someError IO someResult` 292 times in our codebase.
+![image](https://github.com/user-attachments/assets/f405dfca-7d75-404c-b291-3d5140fc8093)
+
+The vast majority of these errors are not used to recover, they are propagated and reported to the user.
+```
+main :: IO ()
+main = toplevelExceptionHandler $ do
+  envCli <- getEnvCli
+  co <- Opt.customExecParser pref (opts envCli)
+  orDie (docToText . renderClientCommandError) $ runClientCommand co
+  
+...
+
+runClientCommand :: ClientCommand -> ExceptT ClientCommandErrors IO ()
+runClientCommand = \case
+  AnyEraCommand cmds ->
+    firstExceptT (CmdError (renderAnyEraCommand cmds)) $ runAnyEraCommand cmds
+  AddressCommand cmds ->
+    firstExceptT AddressCmdError $ runAddressCmds cmds
+  NodeCommands cmds ->
+    runNodeCmds cmds
+      & firstExceptT NodeCmdError
+  ByronCommand cmds ->
+    firstExceptT ByronClientError $ runByronClientCommand cmds
+  CompatibleCommands cmd ->
+    firstExceptT (BackwardCompatibleError (renderAnyCompatibleCommand cmd)) $
+      runAnyCompatibleCommand cmd
+...
+```
+- As a result we have a lot of errors wrapped in errors which makes the code unwieldly and difficult to compose with other code blocks. See image below of incidences of poor composability where we use `firstExceptT` (and sometimes `first`) to wrap errors in other errors.
+
+![image](https://github.com/user-attachments/assets/6a50e8f2-9140-4b7e-afa5-ff778d696742)
+ 
+- The ExceptT IO is a known anti-pattern for these reasons and others as per:
+  - https://github.com/haskell-effectful/effectful/blob/master/transformers.md
+  - https://tech.fpcomplete.com/blog/2016/11/exceptions-best-practices-haskell/
+
+
+# Proposed Solution
+
+Use the RIO monad which is based on the [ReaderT design pattern](https://tech.fpcomplete.com/blog/2017/06/readert-design-pattern/).
+
+### Pros 
+I opt for the `RIO env a` monad instead of `ReaderT e m a` because: 
+- We get additional [logging functionality](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#g:5)
+- Explicit environment dependencies e.g `logError :: HasLogFunc env => Text -> RIO env ()`
+- Better composability i.e no more errors that wrap errors (see above).
```

Comment:
`ReaderT` does not provide error handling so this point seems to be not related?